### DISCUSSION TITLE
[Q&A] Swarm Learning error when num_of_clients > num_of_GPUs

### DISCUSSION THREAD

**Discussion by jingyunc** [NVIDIA/NVFlare#2159] [Q&A] [answered]
### Python version (`python3 -V`)

3.8

### NVFlare version (`python3 -m pip list | grep "nvflare"`)

2.4.0rc3

### Operating system

Ubuntu 20.04

### Have you successfully run any of the following examples?

- [ ] hello-numpy-sag with simulator
- [X] hello-pt with simulator
- [ ] hello-numpy-sag with POC
- [ ] hello-pt with POC

### Please describe your question

On my 4-GPU workstation, I had no problem simulating 2~4 sites with the vanilla Swarm Learning example (https://github.com/NVIDIA/NVFlare/tree/main/examples/advanced/swarm_learning). The command I used was:

`nvflare simulator ./jobs/cifar10_swarm -w /tmp/nvflare/cifar10_swarm -n 2 -t 2`

However, if I set -n to 5 or more in above command, I got following error, regardless of the value of -t:

```
ClientRunner - INFO - [identity=site-1, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=swarm_start, task_id=5750096a-21e8-4494-991f-0e82c1b465d9]: invoking task executor SwarmClientController
ClientRunner - ERROR - [identity=site-1, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=swarm_start, task_id=5750096a-21e8-4494-991f-0e82c1b465d9]: Processing error from executor SwarmClientController: ValueError: invalid model learnable: expect Model type but got <class 'NoneType'>
ClientRunner - ERROR - Traceback (most recent call last):
  File "/home/xxx/venv/flare/lib/python3.8/site-packages/nvflare/private/fed/client/client_runner.py", line 290, in _do_process_task
    reply = executor.execute(task.name, task.data, fl_ctx, abort_signal)
  File "/home/xxx/venv/flare/lib/python3.8/site-packages/nvflare/app_common/ccwf/swarm_client_ctl.py", line 304, in execute
    return super().execute(task_name, shareable, fl_ctx, abort_signal)
  File "/home/xxx/venv/flare/lib/python3.8/site-packages/nvflare/app_common/ccwf/client_ctl.py", line 372, in execute
    initial_model = self.shareable_generator.learnable_to_shareable(learnable, fl_ctx)
  File "/home/xxx/venv/flare/lib/python3.8/site-packages/nvflare/app_common/ccwf/comps/simple_model_shareable_generator.py", line 34, in learnable_to_shareable
    dxo = model_learnable_to_dxo(model_learnable)
  File "/home/xxx/venv/flare/lib/python3.8/site-packages/nvflare/app_common/abstract/model.py", line 69, in model_learnable_to_dxo
  raise ValueError(err)
ValueError: invalid model learnable: expect Model type but got <class 'NoneType'>
```

This is strange because with the Brats18 example (https://github.com/NVIDIA/NVFlare/tree/main/examples/advanced/brats18), I had no problem simulating 8 or 12 sites with the same 4-GPU workstation.

Is there a way to simulate 5 or more sites with the Swarm Learning workflow? Thanks.

**Comment by YuanTingHsieh** [comment]
@jingyunc this is interesting, I just try the master branch and it is working for me.
I have only 1 GPU but I can simulate 5 clients.

Can you paste the exact command you run?
How big is your CPU memory?
How big is your GPU memory?
Can you also check your CPU memory usage and GPU memory usage?

**Reply by jingyunc** [reply]
I just tried simulating 5 clients on 1 GPU on my workstation:

1. Run `nvflare job create -j ./jobs/cifar10_swarm_5site -w swarm_cse_pt_model_learner -sd ../cifar10 -force`
2. Set num_sites = 5 in ./jobs/cifar10_swarm_5site/config_fed_server.json
3. Run `nvflare simulator ./jobs/cifar10_swarm_5site -w ./workspace/cifar10_swarm_5site -n 5 -t 5 -gpu 1 `

It indeed worked.

However when I used 4 GPUs, i.e. `nvflare simulator ./jobs/cifar10_swarm_5site -w ./workspace/cifar10_swarm_5site -n 5 -t 5 -gpu 0,1,2,3`, I still got the same error as above.

My 4 GPU each has 11G memory, and CPU has 128G RAM. During the erroneous run, the GPUs had 0% usage, and CPU used ~9G.

I attached the complete log history, including the error messages. Hope that helps solving the mystery.

[stdout.txt](https://github.com/NVIDIA/NVFlare/files/13433302/stdout.txt)

**Reply by YuanTingHsieh** [reply]
thanks @jingyunc for the logs and information, I will discuss with team and get back to you

**Comment by yhwen** [comment]
Hi @jingyunc. The simulator running multiple clients with multiple GPUs actually is rather complicate here. When running with Simulator, it's not guarantee that all the clients will be running at the same time. Depending on how many clients you run the simulator, some of the clients running could potentially swapped in and swapped out from the active running session. This is depending on the resources of the machine, and the number of clients run you are trying to simulate.  When running with the option: "-n 5 -t 5 -gpu 0,1,2,3", each GPU process will have one active running thread to simulate the client run, and client 1 and client 5 will be alternatively running on the gpu0.  For the cifar10_swarm_5site application, this job requires all the 5 clients actively running at the same time. That's why you saw the run-time error.

If you run the job with another parameters options like "-n 5 -t 4", you will see the same issue for this cifar10_swarm_5site application. If the Job application Does require all the clients are actively running at the same time, like the Aux communication type of applications, you will need to run with the options to ensure all the clients will be actively running at the same time.

However, you may ask why option "-n 5 -t 5 -gpu 1" is able to run this application. This is because in the case of single GPU, we don't force each gpu to only have threads to run the simulation clients. All the 5 clients are running at the same time in 5 threads. Also because this simple particular cifar10_swarm_5site application dees NOT have complicate GPU initialization and resource sharing using multi-threads, so it didn't give the GPU resource conflict errors. But for most of the GPU application, that's not the case. So when specifying the multi-gpu options (>1)in the simulator, we enforce each GPU to be running with threads 1 to avoid the application errors.

**Reply by jingyunc** [reply]
Thank you @yhwen for the clarification. My job application involves 8-12 clients on 4 GPUs, but does not require all clients to run at same time. At each epoch, my workflow goes as follow:

1. Local training on each site, which can be run separately and independently on any available GPUs.
2. After all the local trainings are done, select two sites. The 1st site sends its current local model to the 2nd.
3. The 2nd site merges the incoming model with its current local model, via weighted average.

Since Swarm Learning has some intrinsic limits on site number, should I switch to the Brats18 template instead? I chose the Swarm Learning template because its client-controlled workflow is close to mine. But my application actually trains on Brats dataset. And I had no issue simulating 8-12 sites on 4 GPUs with the Brats18 example job.

Do you think it's feasible to build my application on the Brats18 example, and revise its server-controlled workflow into my client-controlled one, as described above? Or maybe there is a better option? Thanks.

**Reply by jingyunc** [reply]
Also, are the Aux communication options something I could turn on in Swarm Learning? Or is Aux communication an entirely separate application? Thanks.

**Reply by yhwen** [reply]
Hi @jingyunc The key difference of Swarm Learning is the FL control logic are executed on the client sites. Which means, the client controller "distributing" the client tasks to other clients directly. During this tasks distributing, the controller client and task executing client have to be "online" running at the same time. This is different from the server controller type of applications. Because during the simulator run, the server is always running, but the clients could potentially swapped IN and OUT from the execution state, depending on your simulator running parameter options.

For your particular use case, you can consider to use the server controller, for example, ScatterAndGather. The job configuration templates are available in the Brats18 "brats_fedavg" or "brats_central" examples. With this, you should not have problems running 8-12 clients ion 4 GPUs running with the Simulator.

Another option is to choose to run your application in the real world application mode, instead of using the Simulator. In this case, you don't need to change any of your job application set up. You can use the admin tool to submit your job and control your job execution.

The Aux communication is a type of communication directly from one site to another site, either between the server and any clients, or between any clients themselves. You can use this type of communication in any application. But this does not apply to your application right now.

**Reply by jingyunc** [reply]
Thanks @yhwen for the clarification and advice! Indeed, I had no problem running Brats18 example with 8-12 sites on my 4 GPUs.

My concern with Brats18 however is that its workflow (ScatterAndGather) is server-controlled, while my application is supposed to be client-controlled (i.e. peer-to-peer). Before diving into the Brats18 example, I wonder if you could share insights on the following:

1. Do you think it's realistic/feasible for me to move the aggregation in Brtas18 workflow from server-side to the client-side (just like in Swarm Learning)?
2. Theoretically, if the modification in 1. is successful, would the same issue with Swarm Learning rise again when I try to simulate 5+ sites on 4 GPUs? 

Many thanks.

**Reply by yhwen** [reply]
Yes, if you move the controller logic to the client, when you run the application in the Simulator mode, you will face the same issue. This is because the client execution are not guaranteed at the same time. Depending on the number of clients to simulate, and how much the resources you have on the machine, some clients maybe swapped In and Out of the running mode. If it happened for example client A needs to send a task to client B at sometime, and client B has been swapped out at that moment, then the application will fail. Unless you want to make use of the SWAP_IN and SWAP_OUT events, and implement the solution to persist and restore the communication data on the application layer. 

This is a quite complicate solution. By running in the real world application mode using admin tool would be much easier, I believe.

**Reply by jingyunc** [reply]
I see, in that case I will try the real world application mode.

**Reply by jingyunc** [reply]
Quick question: does real life application mode require multiple workstations? For example, if I want to test the cifar10_swarm_5site above, do I need 5 physical workstations? Is it possible to run real life mode on a single workstation? Thanks.

**Reply by yhwen** [reply]
Yes, you can run the real world application mode on a single workstation as well. You can use the CLI command to launch the server and clients at the same time.

**Reply by jingyunc** [reply]
I was able to run cifar10_swarm_5site in real-world mode (thanks to the troubleshoot of #2183 & #2184). However, nvidia-smi showed all 5 sites were using the same GPU. As expected, when running my brats21_swarm_5site job, I got the "cuda out of memory" error.

Is there a place in real-world mode where I could specify GPU availability, just like the "-gpu 0,1,2,3" in simulator mode? Thanks.

**Reply by yhwen** [reply, marked as answer]
You can set the CUDA_VISIBLE_DEVICES OS environment variable to different GPU for each client.

**Reply by jingyunc** [reply]
Confirming cifar10_swarm_5site can run on 4 GPUs, by specifying CUDA_VISIBLE_DEVICES. Thanks.

**Reply by luiji2425** [reply]
I am pretty new to nvflare and I have exactly the same problem running my tf-model on more than 4 GPUs. So simulating 4 clients, with automatic allocation of 1 client per GPU works fine. However, 5 or more clients doesnt work. 

My question: Where and how did you define that CUDA_VISIBLE_DEVICES environment variable? I don't know how and where to implement the answer from @yhwen. Do I need to change something in the source-code for this? Thanks in advance
